# **Waze Project**


Welcome to the Waze Project!

Your Waze data analytics team is still in the early stages of their user churn project. Previously, you were asked to complete a project proposal by your supervisor, May Santner. You have received notice that your project proposal has been approved and that your team has been given access to Waze's user data. To get clear insights, the user data must be inspected and prepared for the upcoming process of exploratory data analysis (EDA).


### **Task 1. Understand the situation**

*   How can you best prepare to understand and organize the provided driver data?


*Begin by exploring your dataset and consider reviewing the Data Dictionary.*

In [1]:
# Import packages 
import pandas as pd
import numpy as np
import matplotlib.pyplot as pt


Then, load the dataset into a dataframe. Creating a dataframe will help you conduct data manipulation, exploratory data analysis (EDA), and statistical activities.



In [2]:
# Load dataset into dataframe
df = pd.read_csv('waze_dataset.csv')

### Summary information**



*Consider the following questions:*

 are there any variables that have missing values?

 what are the data types? How many rows and columns do you have?

Does the dataset have any missing values?

In [3]:

df.head(5)

ID     label  sessions  drives  total_sessions  n_days_after_onboarding  \
0   0  retained       283     226      296.748273                     2276   
1   1  retained       133     107      326.896596                     1225   
2   2  retained       114      95      135.522926                     2651   
3   3  retained        49      40       67.589221                       15   
4   4  retained        84      68      168.247020                     1562   

   total_navigations_fav1  total_navigations_fav2  driven_km_drives  \
0                     208                       0       2628.845068   
1                      19                      64      13715.920550   
2                       0                       0       3059.148818   
3                     322                       7        913.591123   
4                     166                       5       3950.202008   

   duration_minutes_drives  activity_days  driving_days   device  
0              1985.775061             28            19  Android  
1              3160.472914             13            11   iPhone  
2              1610.735904             14             8  Android  
3               587.196542              7             3   iPhone  
4              1219.555924             27            18  Android

In [4]:

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14999 entries, 0 to 14998
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ID                       14999 non-null  int64  
 1   label                    14299 non-null  object 
 2   sessions                 14999 non-null  int64  
 3   drives                   14999 non-null  int64  
 4   total_sessions           14999 non-null  float64
 5   n_days_after_onboarding  14999 non-null  int64  
 6   total_navigations_fav1   14999 non-null  int64  
 7   total_navigations_fav2   14999 non-null  int64  
 8   driven_km_drives         14999 non-null  float64
 9   duration_minutes_drives  14999 non-null  float64
 10  activity_days            14999 non-null  int64  
 11  driving_days             14999 non-null  int64  
 12  device                   14999 non-null  object 
dtypes: float64(3), int64(8), object(2)
memory usage: 1.5+ MB


In [5]:
df.shape

(14999, 13)

1. Lable columns contain null values, it contains only 14299 value where the rest 14999 values

### Null values and summary statistics**

Compare the summary statistics of the 700 rows that are missing labels with summary statistics of the rows that are not missing any values.

**Question:** Is there a discernible difference between the two populations?


In [8]:
#Checking null values
df.isnull().sum()

ID                           0
label                      700
sessions                     0
drives                       0
total_sessions               0
n_days_after_onboarding      0
total_navigations_fav1       0
total_navigations_fav2       0
driven_km_drives             0
duration_minutes_drives      0
activity_days                0
driving_days                 0
device                       0
dtype: int64

In [10]:
# Isolate rows with null values

df_null=df[df.isnull().any(axis=1)]
# Display summary stats of rows with null values

df_null.describe()

ID    sessions      drives  total_sessions  \
count    700.000000  700.000000  700.000000      700.000000   
mean    7405.584286   80.837143   67.798571      198.483348   
std     4306.900234   79.987440   65.271926      140.561715   
min       77.000000    0.000000    0.000000        5.582648   
25%     3744.500000   23.000000   20.000000       94.056340   
50%     7443.000000   56.000000   47.500000      177.255925   
75%    11007.000000  112.250000   94.000000      266.058022   
max    14993.000000  556.000000  445.000000     1076.879741   

       n_days_after_onboarding  total_navigations_fav1  \
count               700.000000              700.000000   
mean               1709.295714              118.717143   
std                1005.306562              156.308140   
min                  16.000000                0.000000   
25%                 869.000000                4.000000   
50%                1650.500000               62.500000   
75%                2508.750000              169.250000   
max                3498.000000             1096.000000   

       total_navigations_fav2  driven_km_drives  duration_minutes_drives  \
count              700.000000        700.000000               700.000000   
mean                30.371429       3935.967029              1795.123358   
std                 46.306984       2443.107121              1419.242246   
min                  0.000000        290.119811                66.588493   
25%                  0.000000       2119.344818               779.009271   
50%                 10.000000       3421.156721              1414.966279   
75%                 43.000000       5166.097373              2443.955404   
max                352.000000      15135.391280              9746.253023   

       activity_days  driving_days  
count     700.000000    700.000000  
mean       15.382857     12.125714  
std         8.772714      7.626373  
min         0.000000      0.000000  
25%         8.000000      6.000000  
50%        15.000000     12.000000  
75%        23.000000     18.000000  
max        31.000000     30.000000

In [11]:
# Isolate rows without null values

df_nonull=df.dropna()
# Display summary stats of rows without null values

df_nonull.describe()

ID      sessions        drives  total_sessions  \
count  14299.000000  14299.000000  14299.000000    14299.000000   
mean    7503.573117     80.623820     67.255822      189.547409   
std     4331.207621     80.736502     65.947295      136.189764   
min        0.000000      0.000000      0.000000        0.220211   
25%     3749.500000     23.000000     20.000000       90.457733   
50%     7504.000000     56.000000     48.000000      158.718571   
75%    11257.500000    111.000000     93.000000      253.540450   
max    14998.000000    743.000000    596.000000     1216.154633   

       n_days_after_onboarding  total_navigations_fav1  \
count             14299.000000            14299.000000   
mean               1751.822505              121.747395   
std                1008.663834              147.713428   
min                   4.000000                0.000000   
25%                 878.500000               10.000000   
50%                1749.000000               71.000000   
75%                2627.500000              178.000000   
max                3500.000000             1236.000000   

       total_navigations_fav2  driven_km_drives  duration_minutes_drives  \
count            14299.000000      14299.000000             14299.000000   
mean                29.638296       4044.401535              1864.199794   
std                 45.350890       2504.977970              1448.005047   
min                  0.000000         60.441250                18.282082   
25%                  0.000000       2217.319909               840.181344   
50%                  9.000000       3496.545617              1479.394387   
75%                 43.000000       5299.972162              2466.928876   
max                415.000000      21183.401890             15851.727160   

       activity_days  driving_days  
count   14299.000000  14299.000000  
mean       15.544653     12.182530  
std         9.016088      7.833835  
min         0.000000      0.000000  
25%         8.000000      5.000000  
50%        16.000000     12.000000  
75%        23.000000     19.000000  
max        31.000000     30.000000

### Null values - device counts**

Next, check the two populations with respect to the `device` variable.

**Question:** How many iPhone users had null values and how many Android users had null values?

In [12]:
#Check the values 
df['device'].value_counts()

iPhone     9672
Android    5327
Name: device, dtype: int64

In [17]:
# Get count of null values by device

df_null['device'].value_counts()

iPhone     447
Android    253
Name: device, dtype: int64

In [22]:
# Calculate % of iPhone nulls and Android nulls

ind=list(df_null['device'].value_counts().index)
val=list(df_null['device'].value_counts().values)


In [26]:
#iphone percentage with null values
total=df_null.shape[0]

In [31]:
#The total number is 700 now find the percentage of iphone 
def percentage(total,device):
    percent=(device/total)*100
    return percent

In [37]:
iphone_perc=percentage(total,val[0])
android_perc=percentage(total,val[1])
print('The percent of iphone devices with null rows: ',"{:.2f}".format(iphone_perc))
print('The percent of android devices with null rows: ',"{:.2f}".format(android_perc))

The percent of iphone devices with null rows:  63.86
The percent of android devices with null rows:  36.14


How does this compare to the device ratio in the full dataset?

In [38]:
# Calculate % of iPhone users and Android users in full dataset

total_df=df.shape[0]
dev=df['device'].value_counts().index
values=df['device'].value_counts().values
iphone=percentage(total_df,values[0])
android=percentage(total_df,values[1])
print('The percent of iphone devices: ',"{:.2f}".format(iphone))
print('The percent of android devices: ',"{:.2f}".format(android))

The percent of iphone devices:  64.48
The percent of android devices:  35.52


The percentage of missing values by each device is consistent with their representation in the data overall.

There is nothing to suggest a non-random cause of the missing data.

Examine the counts and percentages of users who churned vs. those who were retained. How many of each group are represented in the data?

In [45]:
df['label'].value_counts()

retained    11763
churned      2536
Name: label, dtype: int64

In [44]:
# Calculate counts of churned vs. retained


label=df['label'].value_counts().index
label_val=df['label'].value_counts().values
retained=percentage(total_df,label_val[0])
churned=percentage(total_df,label_val[1])
print('The percent of retained users: ',"{:.2f}".format(retained))
print('The percent of churned users: ',"{:.2f}".format(churned))


The percent of retained users:  78.43
The percent of churned users:  16.91


This dataset contains 78% retained users and 16% churned users.

Next, compare the medians of each variable for churned and retained users. The reason for calculating the median and not the mean is that you don't want outliers to unduly affect the portrayal of a typical user. Notice, for example, that the maximum value in the `driven_km_drives` column is 21,183 km. That's more than half the circumference of the earth!

In [53]:
# Average of each column using DataFrame.mean()
df2 = df.median(axis=0)

In [54]:
df2

ID                         7499.000000
sessions                     56.000000
drives                       48.000000
total_sessions              159.568115
n_days_after_onboarding    1741.000000
total_navigations_fav1       71.000000
total_navigations_fav2        9.000000
driven_km_drives           3493.858085
duration_minutes_drives    1478.249859
activity_days                16.000000
driving_days                 12.000000
dtype: float64

In [55]:
# Calculate median values of all columns for churned and retained users
ret=df[df['label']=='retained']
chur=df[df['label']=='churned']

2. The benefit of using the median value of a sample instead of the mean is to avoid the error that arise from outliers

In [57]:
ret_median=ret.median(axis=0)
chur_median=chur.median(axis=0)

In [58]:
ret_median

ID                         7509.000000
sessions                     56.000000
drives                       47.000000
total_sessions              157.586756
n_days_after_onboarding    1843.000000
total_navigations_fav1       68.000000
total_navigations_fav2        9.000000
driven_km_drives           3464.684614
duration_minutes_drives    1458.046141
activity_days                17.000000
driving_days                 14.000000
dtype: float64

In [59]:
chur_median

ID                         7477.500000
sessions                     59.000000
drives                       50.000000
total_sessions              164.339042
n_days_after_onboarding    1321.000000
total_navigations_fav1       84.500000
total_navigations_fav2       11.000000
driven_km_drives           3652.655666
duration_minutes_drives    1607.183785
activity_days                 8.000000
driving_days                  6.000000
dtype: float64

This offers an interesting snapshot of the two groups, churned vs. retained:

Users who churned averaged ~3 more drives in the last month than retained users, but retained users used the app on over twice as many days as churned users in the same time period.

The median churned user drove ~200 more kilometers and 2.5 more hours during the last month than the median retained user.

It seems that churned users had more drives in fewer days, and their trips were farther and longer in duration. Perhaps this is suggestive of a user profile. Continue exploring!

Calculate the median kilometers per drive in the last month for both retained and churned users.

In [62]:
# Group data by `label` and calculate the medians
df_grouped=df.groupby('label').median()



In [64]:
# Divide the median distance by median number of drives
dis_per_drive=df_grouped['driven_km_drives']/df_grouped['drives']

In [65]:
dis_per_drive

label
churned     73.053113
retained    73.716694
dtype: float64

The median user from both groups drove ~73 km/drive. How many kilometers per driving day was this?

In [66]:
# Divide the median distance by median number of driving days
dis_per_day=df_grouped['driven_km_drives']/df_grouped['driving_days']
dis_per_day

label
churned     608.775944
retained    247.477472
dtype: float64

Now, calculate the median number of drives per driving day for each group.

In [67]:
# Divide the median number of drives by median number of driving days
drive_per_day=df_grouped['drives']/df_grouped['driving_days']
drive_per_day

label
churned     8.333333
retained    3.357143
dtype: float64

The median user who churned drove 608 kilometers each day they drove last month, which is almost 250% the per-drive-day distance of retained users. The median churned user had a similarly disproporionate number of drives per drive day compared to retained users.

It is clear from these figures that, regardless of whether a user churned or not, the users represented in this data are serious drivers! It would probably be safe to assume that this data does not represent typical drivers at large. Perhaps the data&mdash;and in particular the sample of churned users&mdash;contains a high proportion of long-haul truckers.

In consideration of how much these users drive, it would be worthwhile to recommend to Waze that they gather more data on these super-drivers. It's possible that the reason for their driving so much is also the reason why the Waze app does not meet their specific set of needs, which may differ from the needs of a more typical driver, such as a commuter.

Finally, examine whether there is an imbalance in how many users churned by device type.

Begin by getting the overall counts of each device type for each group, churned and retained.

In [79]:
# For each label, calculate the number of Android users and iPhone users
df2=df.groupby(['label','device']).size()
df2

label     device 
churned   Android     891
          iPhone     1645
retained  Android    4183
          iPhone     7580
dtype: int64

Now, within each group, churned and retained, calculate what percent was Android and what percent was iPhone.

In [81]:
# For each label, calculate the percentage of Android users and iPhone users
df.groupby('label')['device'].value_counts(normalize=True)

label     device 
churned   iPhone     0.648659
          Android    0.351341
retained  iPhone     0.644393
          Android    0.355607
Name: device, dtype: float64

The ratio of iPhone users and Android users is consistent between the churned group and the retained group, and those ratios are both consistent with the ratio found in the overall dataset.